File: experiments/equations/initial_functions.py
import numpy as np
import torch
from equations.burgers import analytical_burgers_1d 

def gaussian_init(t, x):
    u = np.zeros((t.shape[0], x.shape[0]))
    u[0, :] = np.exp(-(x - 1)**2)
    u[:, 0] =  0
    u[:, -1] = 0

    return u

def burgers_analytical_init(t, x, nu):
    u = np.zeros((t.shape[0], x.shape[0]))

    u_true = analytical_burgers_1d(t[:, None], x[None, :], nu)
    u[0, :] = np.copy(u_true[0, :])
    u[:, 0] =  0 # u_true[:, 0]
    u[:, -1] = 0 # u_true[:, -1]

    return u

def random_init(t, x):
    """
    Random initial conditions
    Statistical analysis and simulation of random shocks in stochastic Burgers equation
    Forcing term f = 0
    """
    #u = np.zeros((t.shape[0], x.shape[0] + 2))
    u = np.zeros((t.shape[0], x.shape[0]))
    nu = np.random.normal(0, 0.25, x.shape[0])
    u[0, :] = np.sin(x) + nu
    u[:, 0] =  0
    u[:, -1] = 0

    return u

def high_dim_random_init(t, x):
    """
    High-dimensional random initial conditions
    Statistical analysis and simulation of random shocks in stochastic Burgers equation
    """
    m = 48
    nu = np.random.normal(0, 1, 2 * m)
    #u = np.zeros((t.shape[0], x.shape[0] + 2))
    u = np.zeros((t.shape[0], x.shape[0]))

    s = [(nu[2 * k - 1] * np.sin(k * x)) + (nu[2 * k - 2] * np.cos(k * x)) for k in range(1, m+1)]

    u[0, :] = (1 / np.sqrt(m)) * np.sum(s, axis=0)
    u[:, 0] =  0
    u[:, -1] = 0

    return u

def analytical_heat_1d(t, x, n=[], c=[], k=1.):
    """
    Analytical solution to 1D heat equation.
    Return solution for a single tuple (t, x)
    @param t : time value
    @param x : space value
    @param c :
    @param n :
    @return u : solution
    @return cn : vector of constant c used for computation.
    """
    L = 1.
    if len(c) == 0:
        c = np.divide(np.random.normal(0., 1., len(n)), n) # n = range(1, n_max), 1 to avoid dividing by 0

    u = np.sum([c[i] * np.exp(- k * (np.pi * n[i] / L)**2 * t) * np.sqrt(2 / L) * np.sin(n[i] * np.pi * x / L) for i in range(len(n))], axis=0)
    return u, c

def heat_analytical_init(t, x, n=[], c=[], k=1.):
    u0 = np.zeros((t.shape[0], x.shape[0]))
    u, _ = analytical_heat_1d(t[:, None], x[None, :], n, c, k)
    u0[0, :] = np.copy(u[0, :])
    u0[:, 0] = 0
    u0[:, -1] = 0

    return u0

File: experiments/julia_src/neural_closure/src/equations/initial_functions.jl
module InitialFunctions

export gaussian_init, random_init, high_dim_random_init, burgers_analytical_init, heat_analytical_init

using LinearAlgebra
using SparseArrays
using OrdinaryDiffEq
using DiffEqFlux
using Random
using Distributions

function gaussian_init(t, x)
  u = zeros(Float64, size(t, 1), size(x, 1))
  u[1, :] .= exp.(-(x .- 1).^2)
  u[:, 1] .= 0
  u[:, end] .= 0
  return u
end

function gaussian2_init(t, x, σ, μ)
  u = zeros(Float64, size(t, 1), size(x, 1))
  u[1, :] .= 1 ./ (σ * sqrt.(2 * pi)) * exp.(-0.5 .* ((x .- μ) ./ σ).^2)
  u[:, 1] .= 0
  u[:, end] .= 0
  return u
end

function random_gaussian_init(t, x, σ, μ)
  d = Normal(1., 0.5)
  σ2 = σ * rand(d)
  μ2 = μ * rand(d)
  return gaussian2_init(t, x, σ2, μ2)
end

function random_init(t, x)
  u = zeros(Float64, size(t, 1), size(x, 1))
  d = Normal(0, .25)
  nu = rand(d, size(x, 1))
  u[1, :] .= sin.(x) + nu
  u[:, 1] .= 0
  u[:, end] .= 0

  return u
end

function high_dim_random_init(t, x, m=28)
  d = Normal(0., 1.)
  nu = rand(d, 2 * m)
  s = [nu[2 * k] * sin.(k * x) + nu[2 * k - 1] * cos.(k * x) for k in range(1, m, step=1)]

  u0 = zeros(Float64, size(t, 1), size(x, 1))
  u0[1, :] .= (1 / sqrt(m)) .* sum(s)
  u0[:, 1] .= 0
  u0[:, end] .= 0

  return u0
end

function high_dim_random_init2(t, x, m=28)
  d = Normal(0., 1.)
  nu = rand(d, 2 * m)
  x = (x .- x[1]) ./ (x[end] .- x[1])
  s = [nu[2 * k] * sin.(2 * pi * k * x) + nu[2 * k - 1] * cos.(2 * pi * k * x) for k in range(1, m, step=1)]

  u0 = zeros(Float64, size(t, 1), size(x, 1))
  u0[1, :] .= (1 / sqrt(m)) .* sum(s)

  return u0
end


function analytical_burgers_1d(t, x, nu)
end

function burgers_analytical_init(t, x, nu)
  u0 = zeros(Float64, size(t, 1), size(x, 1))
  # u = analytical_burgers_1d(t, x, nu)
  # u0[1, :] .= u[1, :]
  # u0[:, 1] .= 0
  # u0[:, end] .= 0
  return u0
end

function advecting_shock(t, x, nu)
  Re = 1. / nu;
  t0 = exp(Re / 8.);

  u0 = zeros(Float64, size(t, 1), size(x, 1));
  u0[1, :] .= x ./ (1 .+ sqrt(1. / t0) * exp.(Re * (x.^2 ./ 4)));
  u0[:, 1] .= 0
  u0[:, end] .= 0

  return u0
end

function analytical_heat_1d(t, x, n=[], c=[], κ=1.)
  xₘᵢₙ, xₘₐₓ = x[1], x[end];
  L = xₘₐₓ - xₘᵢₙ;
  N = size(n)[1];
  k = 1:N;

  if (size(c, 1) == 0)
    c = Array(randn(N) ./ k);  # n = range(1, n_max, step=1), 1 to avoid division by 0
  end

  aₖ(k, x) = sqrt(2 / L) * sin(π * k * (x - xₘᵢₙ) / L);
  u(t, x) = sum(cₖ * exp(- κ * π^2 * k^2 / L * t) * aₖ(k, x) for (cₖ, k) in zip(c, k))
  return Array([u(a, b) for a in t, b in x]) 
end


function heat_analytical_init(t, x, n=[], c=[], κ=1.)
  u₀ = zeros(Float64, size(t, 1), size(x, 1))
  u = analytical_heat_1d([t[1]], x, n, c, κ)
  u₀[1, :] = copy(u[1, :])
  u₀[:, 1] .= 0
  u₀[:, end] .= 0

  return u₀
end

end
